### PR TITLE
Add migration to send users message to reconnect on plugin startup

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR aims to fix the revoked token issue by doing the following on startup:

- acquire mutex to gain control, to run a data migration
- for all connected users:
  - remove stored token, and any other gitlab-related information, from the kv store
  - DM the user telling them to reconnect their account

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-gitlab/issues/261

### CODE REVIEW

File: server/oauth_migration.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+func (p *Plugin) checkAndPerformOAuthTokenMigration() error {
+	mutexAPI := cluster.MutexPluginAPI(p.API)
+	mutex, err := cluster.NewMutex(mutexAPI, oauthMigrationMutexKey)
+	if err != nil {
+		return err
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
```
Comment (mickmister) [inline server/oauth_migration.go:28]: In the first commit of the PR, I had this call further down, and not `defer`ed. I think it should be `defer`ed up here so we can make sure the migration completes before any of the plugins are considered "activated". This way we can avoid the thundering hurd problem of several browsers connecting at once.

File: server/oauth_migration.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Migration is in progress or already completed
+	if status != "" {
+		return nil
+	}
+
+	_, _ = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusInProgress)
```
Comment (spirosoik) [inline server/oauth_migration.go:41]: nit: should we check the err here?

**Review by spirosoik — COMMENTED** [review]
minor comment

File: server/oauth_migration.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		}
+
+		keysToAdd := []string{}
+		for _, key := range keys {
+			if strings.HasSuffix(key, GitlabTokenKey) {
+				keysToAdd = append(keysToAdd, key)
+			}
+
+			if strings.HasSuffix(key, GitlabUsernameKey) || strings.HasSuffix(key, GitlabIDUsernameKey) {
+				_ = p.client.KV.Delete(key)
```
Comment (mickmister) [inline server/oauth_migration.go:74]: I actually don't think we should delete these keys. Incoming webhooks can still successfully show the connected mentions this way, before the user reconnects their account.

File: server/oauth_migration.go (hunk 1 above)
Comment (mickmister) [reply server/oauth_migration.go:28]: Note that this means all plugin instances are blocking on `OnActivate` until the migration is complete. This will slow down the start-up time of the server. I wonder how we can make this more asynchronous. We have control over the frontend of the plugin so we can make changes there if necessary.

File: server/oauth_migration.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		_, kvErr := p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusError)
+		if kvErr != nil {
+			p.client.Log.Warn("error setting migration status to "+oauthMigrationStatusError, "error", err.Error())
+		}
+
+		return err
+	}
+
+	_, err = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusComplete)
```
Comment (mickmister) [inline server/oauth_migration.go:56]: It would be much better to have a generic, consistent implementation of this to use among plugins

File: server/oauth_migration.go (hunk 2 above)
Comment (mickmister) [reply server/oauth_migration.go]: @spirosoik Logging those errors now. Please take another look if you'd like

File: server/oauth_migration.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		return nil
+	}
+
+	_, err = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusInProgress)
+	if err != nil {
+		p.client.Log.Warn("error setting migration status to "+oauthMigrationStatusInProgress, "error", err.Error())
+	}
+
+	err = p.notifyAllConnectedUsersToReconnect()
+	if err != nil {
```
Comment (m1lt0n) [inline server/oauth_migration.go:47]: (question) `notifyAllConnectedUsersToReconnect` returns an error only on listing failing. Since we're counting the errors per user, should this also fail if `numErrors` matches `allKeys` length (ie it failed for every single user), as it may indicate some issue with the migration? Or would it be an overkill/edge case?

**Review by m1lt0n — APPROVED** [review]
Great PR! Love the tests, too. Left a minor comment/question.

File: server/oauth_migration.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+		keysToAdd := []string{}
+		for _, key := range keys {
+			if strings.HasSuffix(key, GitlabTokenKey) {
+				keysToAdd = append(keysToAdd, key)
+			}
+		}
+
+		allKeys = append(allKeys, keysToAdd...)
+		page++
```
Comment (m1lt0n) [inline server/oauth_migration.go:85]: (suggestion)(non-blocking) We may save an extra request that's always done (for page 1) if we break if `len(keys)<100`

**Review by hanzei — CHANGES_REQUESTED** [review]
Why did you take a rather radical approach of disconnecting all users vs. promoting them to reconnect when ever a GitLab API request failed with an authentication problem?

**Comment by mickmister** [comment]
@hanzei I want to get this fixed and released, and I feel that it's taking too long to get the more complex solution merged. I don't know how we can get that across the finish line given our current priorities, so I made this PR as an alternative.

How is this for a slightly different solution:

- We ping users om startup just as this PR does
- We don't disconnect users
- We make it so we don't log the revoked token issue, or just make sure we log it as `DEBUG` instead of `ERROR`
